sycl/include/sycl/esimd/detail/esimd_util.hpp

```diff
@@ -0,0 +1,233 @@
+//==----------------- esimd_util.hpp - DPC++ Explicit SIMD API -------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Utility functions used for implementing Explicit SIMD APIs.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/esimd/detail/esimd_types.hpp>
+
+namespace __esimd {
+
+/// Constant in number of bytes.
+enum { BYTE = 1, WORD = 2, DWORD = 4, QWORD = 8, OWORD = 16, GRF = 32 };
+
+/// Compute the next power of 2 at compile time.
+static ESIMD_INLINE constexpr unsigned int getNextPowerOf2(unsigned int n,
+                                                           unsigned int k = 1) {
+  return (k >= n) ? k : getNextPowerOf2(n, k * 2);
+}
+
+/// Check if a given 32 bit positive integer is a power of 2 at compile time.
+static ESIMD_INLINE constexpr bool isPowerOf2(unsigned int n) {
+  return (n & (n - 1)) == 0;
+}
+
+static ESIMD_INLINE constexpr bool isPowerOf2(unsigned int n,
+                                              unsigned int limit) {
+  return (n & (n - 1)) == 0 && n <= limit;
+}
+
+static ESIMD_INLINE constexpr unsigned log2(unsigned n) {
+  return (n > 1) ? 1 + log2(n >> 1) : 0;
+}
+
+} // namespace __esimd
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace intel {
+namespace gpu {
+
+namespace details {
+
+/// type traits
+template <typename T> struct is_esimd_vector {
+  static const bool value = false;
+};
+template <typename T, int N>
+struct is_esimd_vector<sycl::intel::gpu::simd<T, N>> {
+  static const bool value = true;
+};
+template <typename T, int N>
+struct is_esimd_vector<sycl::intel::gpu::vector_type<T, N>> {
+  static const bool value = true;
+};
+
+template <typename T>
+struct is_esimd_scalar
+    : std::integral_constant<
```

Comment:
Is this the same as `std::is_arithmetic<T>::value`?

Reply:
Need to return true for fp16 when we get fp16 working

Reply:
Then you can use `detail::is_arithmetic<T>::value` instead: https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/type_traits.hpp#L193

Reply:
resolved